README.md
# Book Store (Manually RESTified)

<link rel="stylesheet" type="text/css" media="all" href="markdown/figure.css" />

Manually RESTified server side java code of a fake Book-Store for academic case studies.

![version](https://img.shields.io/badge/version-1.0-brightgreen)
![coverage](https://img.shields.io/badge/coverage-91%25-brightgreen)
![building](https://img.shields.io/badge/build-passing-brightgreen)

## About

This is a modified version of the [Bookstore Internals](https://github.com/kartoffelquadrat/BookStoreInternals). Bookstore functionality has been exposed through a REST interface.

### Changes

 * Condensed changelog
   * [Launcher](markdown/changelog.md#launcher)
   * [Spring REST Annotations](markdown/changelog.md#annotations)
   * [Maven ```pom.xml```](markdown/changelog.md#maven)
 * Git changelog (Pre-RESTify commit <-> Post-RESTify commit):  
[```git diff d45b98..8a801```](markdown/restify-diff.txt)
 * [Statistics](markdown/stats.md) of code modifications.


### REST interface layout

 * [Full API specification as ARC](markdown/bookstore.arc)
 * Resource layout visualization:  
![RTL](markdown/RTL-bookstore.png#figure)

 > Note: For comparability to TouchCORE's *RESTify* concern, this implementation contains no **query string parameters** (e.g.: ```http://url?name=value```, ```@RequestParam String name```)  
Instead of a (optional for G, mandatory for Po) request parameter ```isbn``` (to filter on the stocklocation resource), an additional subresource has been added. This design is disputable, but pragmatic for the purpose if this demo application.

## Usage

On execution, the bookstore will simply print a listing of persisted initial dummy data.

 1. Get the sources:  
```git clone https://github.com/kartoffelquadrat/BookStoreManualRestify.git```

 2. Compile the project with maven. This provides you with a self contained ```jar``` in the ```target```folder:  
```mvn clean package```
 
 3. Run the jar:  
 ```java -jar target/BookStore.jar```

## Contact / Pull Requests

 * Author: Maximilian Schiedermeier ![email](markdown/email.png)
 * Github: Kartoffelquadrat
 * Webpage: https://www.cs.mcgill.ca/~mschie3
 * License: [MIT](https://opensource.org/licenses/MIT)



markdown/changelog.md
# Changelog

List of source code modifications, compared to the [original BookStore](https://github.com/kartoffelquadrat/BookStoreInternals).

## Launcher

New launcher added:

```java
package eu.kartoffelquadrat.bookstoreinternals;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

/**
 * This class powers up Spring and ensures the annotated controllers are detected.
 */
@SpringBootApplication
public class RestLauncher {
    public static void main(String[] args) {

        SpringApplication.run(RestLauncher.class, args);
    }
}
```

## Annotations

Search for annotations in RESTified source code. (Non spring annotations filtered)

```bash
$ grep -nre "@" src | grep -v "author" | grep -v "param" | grep -v "return" | grep -v "Override" | grep -v "Test" | cut -f3 -d":" | sed 's/\ \ \ \ //' | sort
@DeleteMapping("/bookstore/isbns/{isbn}/comments")
@DeleteMapping("/bookstore/isbns/{isbn}/comments/{commentid}")
@GetMapping("/bookstore/isbns")
@GetMapping("/bookstore/isbns/{isbn}")
@GetMapping("/bookstore/isbns/{isbn}/comments")
@GetMapping("/bookstore/stocklocations")
@GetMapping("/bookstore/stocklocations/{stocklocation}")
@GetMapping("/bookstore/stocklocations/{stocklocation}/{isbn}")
@PostMapping("/bookstore/isbns/{isbn}/comments")
@PostMapping("/bookstore/isbns/{isbn}/comments/{commentid}")
@PostMapping("/bookstore/stocklocations/{stocklocation}/{isbn}")
@PutMapping("/bookstore/isbns/{isbn}")
@RestController
@RestController
@RestController
@SpringBootApplication
public BookDetailsImpl getBookDetails(@PathVariable("isbn") Long isbn) {
public Map<Long, Integer> getEntireStoreStock(@PathVariable("stocklocation") String city) {
public Map<Long, String> getAllCommentsForBook(@PathVariable("isbn") long isbn) {
public int getStock(@PathVariable("stocklocation") String city, @PathVariable("isbn") Long isbn) {
public void addBookToAssortment(@RequestBody BookDetailsImpl bookDetails) {
public void addComment(@PathVariable("isbn") long isbn, @RequestBody String comment) {
public void deleteComment(@PathVariable("isbn") long isbn,@PathVariable("commentid") long commentId) {
public void editComment(@PathVariable("isbn") long isbn, @PathVariable("commentid") long commentId, @RequestBody  String updatedComment) {
public void removeAllCommentsForBook(@PathVariable("isbn") long isbn) {
public void setStock(@PathVariable("stocklocation")String city, @PathVariable("isbn") Long isbn, @RequestBody Integer amount) {
```

## Maven

 * Name modified:  
```<name>bookstoremanualrest</name>```
 * Parent added:  
```xml
    <!-- this parent block lets this maven project extend a prepared spring specific template.-->
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.3.0.RELEASE</version>
    </parent>
```
 * Legacy build plugin removed:  
```xml
            <!-- specify main class for exec goal -->
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>exec-maven-plugin</artifactId>
                <version>1.6.0</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>java</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <mainClass>eu.kartoffelquadrat.bookstoreinternals.DesktopLauncher</mainClass>
                </configuration>
            </plugin>

            <!-- specify main class for JAR manifest-->
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>3.2.0</version>
                <configuration>
                    <archive>
                        <manifest>
                            <mainClass>eu.kartoffelquadrat.bookstoreinternals.DesktopLauncher</mainClass>
                        </manifest>
                    </archive>
                </configuration>
            </plugin>
```
 * Spring build plugin added:  
```xml
	    <!-- Spring specific build plugin, produces self contained JAR with default launcher class.-->
	    <plugin>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-maven-plugin</artifactId>
		<configuration>
		    <fork>true</fork>
		    <mainClass>eu.kartoffelquadrat.bookstoreinternals.RestLauncher</mainClass>
		</configuration>
		<executions>
		    <execution>
			<goals>
			    <goal>repackage</goal>
			</goals>
		    </execution>
		</executions>
	    </plugin>
```


src/main/java/eu/kartoffelquadrat/bookstoreinternals/CommentsImpl.java
package eu.kartoffelquadrat.bookstoreinternals;

import org.springframework.web.bind.annotation.*;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * DAO for comments on books. Singleton.
 *
 * @author Maximilian Schiedermeier
 */
@RestController
public class CommentsImpl implements Comments {

    private static Comments singletonReference = null;
    Map<Long, Map<Long, String>> commentsByIsbn;

    /**
     * Public default constructor for frameworks that automatically initialize beans as singletons.
     */
    public CommentsImpl() {
        this(true);
    }

    /**
     * Private constructor for singleton pattern.
     *
     * @param populate flag causes the comments to initialized with dummy data, if set to true
     */
    private CommentsImpl(boolean populate) {

        commentsByIsbn = new LinkedHashMap<>();
        if (populate)
            populateWithDummyData();
    }

    public static Comments getInstance() {
        if (singletonReference == null)
            singletonReference = new CommentsImpl(true);

        return singletonReference;
    }

    @GetMapping("/bookstore/isbns/{isbn}/comments")
    public Map<Long, String> getAllCommentsForBook(@PathVariable("isbn") long isbn) {

        // If there are no comments yet, return new empty map.
        if (!commentsByIsbn.containsKey(isbn))
            return new LinkedHashMap<Long, String>();

        return Collections.unmodifiableMap(
                commentsByIsbn.get(isbn));
    }

    /*
     * Although comments are indexed by IDs, the creation of a new comment does not require an ID, for the ID is a
     * dynamic / non-global / internal ID.
     */
    @PostMapping("/bookstore/isbns/{isbn}/comments")
    public void addComment(@PathVariable("isbn") long isbn, @RequestBody String comment) {

        // Verify the comment is not empty
        if (comment.trim().isEmpty())
            throw new RuntimeException("Empty comments can not be added.");

        // Verify that the isbn is valid (indexed by assortment)
        if (!AssortmentImpl.getInstance().getEntireAssortment().contains(isbn))
            throw new RuntimeException("Comment can not be added. No such isbn in assortment: " + isbn);

        // Create new map for isbn, if this is the first comment
        if (!commentsByIsbn.containsKey(isbn))
            commentsByIsbn.put(isbn, new LinkedHashMap<Long, String>());

        // Actually add the comment
        commentsByIsbn.get(isbn).put(generateCommentId(isbn), comment);
    }

    @DeleteMapping("/bookstore/isbns/{isbn}/comments/{commentid}")
    public void deleteComment(@PathVariable("isbn") long isbn,@PathVariable("commentid") long commentId) {

        if (!commentsByIsbn.containsKey(isbn))
            throw new RuntimeException("Comment can not be removed. No such isbn in assortment: " + isbn);

        if (!commentsByIsbn.get(isbn).containsKey(commentId))
            throw new RuntimeException("Comment can not be removed. For this book there is no comment with ID: " + commentId);

        commentsByIsbn.get(isbn).remove(commentId);
    }

    @DeleteMapping("/bookstore/isbns/{isbn}/comments")
    public void removeAllCommentsForBook(@PathVariable("isbn") long isbn) {

        if (!commentsByIsbn.containsKey(isbn))
            throw new RuntimeException("Comments can not be removed. No such isbn in assortment: " + isbn);

        commentsByIsbn.remove(isbn);
    }

    @PostMapping("/bookstore/isbns/{isbn}/comments/{commentid}")
    public void editComment(@PathVariable("isbn") long isbn, @PathVariable("commentid") long commentId, @RequestBody  String updatedComment) {

        // Verify the comment is not empty
        if (updatedComment.trim().isEmpty())
            throw new RuntimeException("Empty comments are not allowed.");

        if (!commentsByIsbn.containsKey(isbn))
            throw new RuntimeException("Comment can not be altered. No such isbn in assortment: " + isbn);

        if (!commentsByIsbn.get(isbn).containsKey(commentId))
            throw new RuntimeException("Comment can not be altered. For this book there is no comment with ID: " + commentId);

        commentsByIsbn.get(isbn).put(commentId, updatedComment);
    }

    /**
     * Adds fake comments to some indexed books.
     */
    private void populateWithDummyData() {

        addComment(Long.valueOf("9780739360385"), "Amazing book!");
        addComment(Long.valueOf("9780739360385"), "A must read!");
        addComment(Long.valueOf("9780553382563"), "A classic!");
        addComment(Long.valueOf("9780553382563"), "Would read it again.");
        addComment(Long.valueOf("9781977791122"), "Love it!");
        addComment(Long.valueOf("9780262538473"), "Fantastic!");
    }

    /**
     * Creates a unique but random comment id, for a given book.
     *
     * @param isbn as the id of the book to receive the comment.
     * @return a random long that is not yet used as a comment id for that book.
     */
    private long generateCommentId(long isbn) {
        long randomLong = (long) (Math.random() * 1000000);

        if (commentsByIsbn.get(isbn).containsKey(randomLong))
            return generateCommentId(isbn);
        else
            return randomLong;
    }

    public String toString()
    {
        StringBuilder sb = new StringBuilder("\n **************\n * Comments * \n **************\n");
        for(long isbn: commentsByIsbn.keySet())
        {
            sb.append(isbn+":\n");
            for(long commentId : commentsByIsbn.get(isbn).keySet())
            {
                sb.append(" > "+commentId + ": "+commentsByIsbn.get(isbn).get(commentId)+"\n");
            }
        }
        return sb.toString();
    }
}


src/main/java/eu/kartoffelquadrat/bookstoreinternals/AssortmentImpl.java
package eu.kartoffelquadrat.bookstoreinternals;

import org.springframework.web.bind.annotation.*;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * @author Maximilian Schiedermeier
 */
@RestController
public class AssortmentImpl implements Assortment {

    private static Assortment singletonReference = null;
    Map<Long, BookDetailsImpl> assortmentMap;

    /**
     * Public default constructor for frameworks that automatically initialize beans as singletons.
     */
    public AssortmentImpl() {

        this(true);
    }

    /**
     * Default constructor adds dummy data to the "in-ram DB"
     *
     * @param populate flag causes the assortment to be initialized with dummy data, if set to true
     */
    private AssortmentImpl(boolean populate) {

        assortmentMap = new LinkedHashMap<Long, BookDetailsImpl>();

        if (populate) {
            populateWithDummyData();
        }
    }

    public static Assortment getInstance() {
        if (singletonReference == null)
            singletonReference = new AssortmentImpl(true);

        return singletonReference;
    }

    @GetMapping("/bookstore/isbns")
    public Collection<Long> getEntireAssortment() {
        return assortmentMap.keySet();
    }

    @GetMapping("/bookstore/isbns/{isbn}")
    public BookDetailsImpl getBookDetails(@PathVariable("isbn") Long isbn) {
        return assortmentMap.get(isbn);
    }

    @Override
    @PutMapping("/bookstore/isbns/{isbn}")
    public void addBookToAssortment(@RequestBody BookDetailsImpl bookDetails) {
        if(assortmentMap.containsKey(bookDetails.getIsbn()))
            throw new RuntimeException("Book can not be indexed, the ISBN conflicts to an existing book.");

        assortmentMap.put(bookDetails.getIsbn(), bookDetails);
    }

    /**
     * Adds fake books to the assortmentMap, adds dummy cities with dummy stocks.
     */
    private void populateWithDummyData() {

        assortmentMap.put(Long.valueOf("9780739360385"), new BookDetailsImpl(Long.valueOf("9780739360385"), "Harry Potter and the deathly hallows", "J.K.Rowling", 5450, "Harry has finally come of age, and finally started on his final journey to defeat Voldemort for good. The Durselys are forced to go into hiding so that Voldemort’s Death Eaters will not torture them for information, and Harry sets off with Ron and Hermione on a difficult quest to find and destroy the last of Voldemort’s Horcruxes. Only once those have been destroyed, Harry knows, can Voldemort truly be killed."));
        assortmentMap.put(Long.valueOf("9780553382563"), new BookDetailsImpl(Long.valueOf("9780553382563"), "I, Robot", "Isaac Asimov", 2300, "Can you imagine living in a world where the sight of human-like robots roaming the streets isn’t unusual? That might be a world we’re going to have to adapt to soon. Science-fiction author Isaac Asimov expresses the benefits and flaws of such a fascinating world in his 1942 short story collection I, Robot."));
        assortmentMap.put(Long.valueOf("9781977791122"), new BookDetailsImpl(Long.valueOf("9781977791122"), "White nights", "Fyodor Dostoevsky", 845, "Set in 1840s St. Petersburg, the short story White Nights starts with its unnamed narrator walking around the streets of the city by himself. He states that though he has lived in the city eight years, he has only acquaintances, including an old man he frequently meets on one of his walks but never talks to."));
        assortmentMap.put(Long.valueOf("9780262538473"), new BookDetailsImpl(Long.valueOf("9780262538473"), "The invincible", "Stanislaw Lem", 2327, "In the grand tradition of H. G. Wells and Jules Verne, Stanislaw Lem's The Invincible tells the story of a space cruiser sent to an obscure planet to determine the fate of a sister spaceship whose communication with Earth has abruptly ceased. Landing on the planet Regis III, navigator Rohan and his crew discover a form of life that has apparently evolved from autonomous, self-replicating machines―perhaps the survivors of a “robot war.” Rohan and his men are forced to confront the classic quandary: what course of action can humanity take once it has reached the limits of its knowledge? In The Invincible, Lem has his characters confront the inexplicable and the bizarre: the problem that lies just beyond analytical reach."));

    }

    @Override
    public String toString()
    {
        StringBuilder sb = new StringBuilder("\n **************\n * Assortment * \n **************\n");
        for(BookDetailsImpl details: assortmentMap.values())
        {
            sb.append(details);
        }
        return sb.toString();
    }
}


src/main/java/eu/kartoffelquadrat/bookstoreinternals/GlobalStockImpl.java
package eu.kartoffelquadrat.bookstoreinternals;

import org.springframework.web.bind.annotation.*;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * @author Maximilian Schiedermeier
 */
@RestController
public class GlobalStockImpl implements GlobalStock {

    private static GlobalStock singletonReference;
    Map<String, LocalStock> stocksPerCity;

    /**
     * Public default constructor for frameworks that automatically initialize beans as singletons.
     */
    public GlobalStockImpl() {

        this(true);
    }


    /**
     * Default constructor adds dummy data to the "in-ram DB"
     *
     * @param populate flag causes the global stock to be initialized with dummy data, if set to true
     */
    private GlobalStockImpl(boolean populate) {
        stocksPerCity = new LinkedHashMap<>();

        if (populate) {
            populateWithDummyData();
        }
    }

    /**
     * Singleton access method.
     *
     * @return the singleton instance.
     */
    public static GlobalStock getInstance() {

        if (singletonReference == null)
            singletonReference = new GlobalStockImpl(true);

        return singletonReference;
    }

    @GetMapping("/bookstore/stocklocations/{stocklocation}/{isbn}")
    public int getStock(@PathVariable("stocklocation") String city, @PathVariable("isbn") Long isbn) {

        if (!stocksPerCity.containsKey(city))
            throw new RuntimeException("Can not lookup amount in stock. No such city: " + city);

        return stocksPerCity.get(city).getAmount(isbn);
    }

    @PostMapping("/bookstore/stocklocations/{stocklocation}/{isbn}")
    public void setStock(@PathVariable("stocklocation")String city, @PathVariable("isbn") Long isbn, @RequestBody Integer amount) {

        if (!stocksPerCity.containsKey(city))
            throw new RuntimeException("Can not update amount in stock. No such city: " + city);

        stocksPerCity.get(city).setAmount(isbn, amount);
    }

    @Override
    @GetMapping("/bookstore/stocklocations")
    public Collection<String> getStoreLocations() {
        return stocksPerCity.keySet();
    }

    @Override
    @GetMapping("/bookstore/stocklocations/{stocklocation}")
    public Map<Long, Integer> getEntireStoreStock(@PathVariable("stocklocation") String city) {
        return stocksPerCity.get(city).getEntireStock();
    }

    private void populateWithDummyData() {

        // Add some locations.
        stocksPerCity.put("Montréal", new LocalStockImpl());
        stocksPerCity.put("München", new LocalStockImpl());
        stocksPerCity.put("Osterhofen", new LocalStockImpl());
        stocksPerCity.put("Lyon", new LocalStockImpl());

        // initialize stock for locations.
        stocksPerCity.get("Montréal").setAmount(Long.valueOf("9780739360385"), 1);
        stocksPerCity.get("Montréal").setAmount(Long.valueOf("9781977791122"), 2);
        stocksPerCity.get("Montréal").setAmount(Long.valueOf("9780262538473"), 3);

        stocksPerCity.get("München").setAmount(Long.valueOf("9780739360385"), 50);
        stocksPerCity.get("München").setAmount(Long.valueOf("9780553382563"), 2);
        stocksPerCity.get("München").setAmount(Long.valueOf("9781977791122"), 7);

        stocksPerCity.get("Osterhofen").setAmount(Long.valueOf("9780739360385"), 15);
        stocksPerCity.get("Osterhofen").setAmount(Long.valueOf("9780553382563"), 2);
        stocksPerCity.get("Osterhofen").setAmount(Long.valueOf("9781977791122"), 5);
        stocksPerCity.get("Osterhofen").setAmount(Long.valueOf("9780262538473"), 8);

        stocksPerCity.get("Lyon").setAmount(Long.valueOf("9780739360385"), 4);
        stocksPerCity.get("Lyon").setAmount(Long.valueOf("9780262538473"), 2);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("\n **************\n *    Stock   * \n **************\n");
        for (String city : stocksPerCity.keySet())
        {
            sb.append(city+":\n");
            sb.append(stocksPerCity.get(city));
        }
        return sb.toString();
    }

}
